flake8_trio.py

```diff
@@ -34,29 +33,68 @@ def is_trio_call(node: ast.AST, *names: str) -> Optional[str]:
 class Visitor(ast.NodeVisitor):
     def __init__(self) -> None:
         super().__init__()
-        self.problems: list[tuple[int, int]] = []
+        self.problems: List[Error] = []
 
     def visit_With(self, node: ast.With) -> None:
-        for item in node.items:
-            call = is_trio_call(item.context_expr, "fail_after", "move_on_after")
-            if call and not any(isinstance(x, ast.Await) for x in ast.walk(node)):
-                self.problems.append(
-                    (item.lineno, item.col_offset, TRIO100.format(call))
-                )
+        self.check_for_trio100(node)
+        self.generic_visit(node)
 
-        # Don't forget to visit the child nodes for other errors!
+    def visit_AsyncWith(self, node: ast.AsyncWith) -> None:
+        self.check_for_trio100(node)
         self.generic_visit(node)
 
+    def check_for_trio100(self, node: Union[ast.With, ast.AsyncWith]) -> None:
+        # Context manager with no `await` call within
+        for item in (i.context_expr for i in node.items):
+            call = is_trio_call(item, "fail_after", "move_on_after")
+            if call and not any(isinstance(x, ast.Await) for x in ast.walk(node)):
+                self.problems.append(TRIO100(item.lineno, item.col_offset, call))
+
 
 class Plugin:
     name = __name__
     version = __version__
 
-    def __init__(self, tree: ast.AST) -> None:
-        self._tree = tree
+    def __init__(
+        self, tree: Optional[ast.AST] = None, filename: Optional[str] = None
+    ) -> None:
+        if tree is None:
+            assert filename is not None
+            self._tree = self.load_file(filename)
+        else:
+            self._tree = tree
+
+    def load_file(self, filename: str) -> ast.AST:
+        """Loads the file in a way that auto-detects source encoding.
 
-    def run(self) -> Generator[tuple[int, int, str, Type[Any]], None, None]:
+        Stolen from flake8_import_order because it's good.
+        """
+        lines = pycodestyle.readlines(filename)
+        return ast.parse("".join(lines))
+
+    def run(self) -> Generator[Tuple[int, int, str, Type[Any]], None, None]:
         visitor = Visitor()
         visitor.visit(self._tree)
-        for line, col, message in visitor.problems:
-            yield line, col, message, type(self)
+        for problem in visitor.problems:
+            yield problem.values()
+
+
+class Error:
+    def __init__(self, lineno: int = -1, col: int = -1, message: str = ""):
+        self.lineno = lineno
+        self.col = col
+        self.message = message
+        self.err_type = type(Plugin)
+
+    def values(self):
+        return (self.lineno, self.col, self.message, self.err_type)
+
+
+class TRIO100(Error):
```

Comment:
I don't think this is worth the overhead when we're just going to work with tuples anyway.

If the iteration I had was unsatisfying, how about we include `Plugin` in each tuple in `Visitor.problems`, so that we can just `yield from visitor.problems`?

Reply:
It was because I was hitting typing problems with the `partial` solution that bugbear had, but yeah I simplified it down a lot and am just doing a tuple + factory function.

Lots of minor variations this could take, could do a `namedtuple` (but atm never accesses any members anyway so w/e), or could do away with the function and let callers do the format + run add the type.